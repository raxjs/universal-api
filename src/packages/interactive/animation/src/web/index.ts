import { Animation, AnimationOptions } from '../types';
import AnimationImpl from './AnimationImpl';

export function createAnimation(options?: AnimationOptions): Animation {
  return new AnimationImpl(options);
}

export default {
  createAnimation,
};
