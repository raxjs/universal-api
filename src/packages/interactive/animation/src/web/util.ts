import { AnimationAction, AnimationOptions } from '../types';
import { normalizeOptions } from '../common';
import { CONTAINER_NAME } from '@utils/constant';

/**
 * normalize unit
 * @param value
 * @param unit
 * @param force
 */
export function normalizeUnit(value: number | string, unit: string, force = false): string {
  if (force || typeof value === 'number') {
    return `${value}${unit}`;
  }
  return value;
}

/**
 * merge default options
 * @param options
 */
export function getDefaultOptions(options?: AnimationOptions): AnimationOptions {
  return normalizeOptions(
    {
      duration: 400,
      timingFunction: 'linear',
      delay: 0,
      transformOrigin: '50% 50% 0',
      ...options,
    },
    CONTAINER_NAME.WEB,
  );
}

/**
 * handle animation actions queue
 * @param actions
 * @param fn
 * @param onFinish
 */
export function handleActionsQueue(
  actions: AnimationAction[],
  fn: (action: AnimationAction, callback: () => void) => any,
  onFinish?: () => any,
) {
  actions = actions.slice();
  const action = actions.shift();
  if (action) {
    fn(action, () => handleActionsQueue(actions, fn, onFinish));
  } else if (onFinish) {
    onFinish();
  }
}
