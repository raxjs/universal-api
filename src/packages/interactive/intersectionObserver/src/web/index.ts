import { CreateIntersectionObserver, OPTIONS, RECT, CallbackOptions } from '../types';

function buildMarginStr(margins: RECT): string {
  return `${Number(margins.top) || 0 }px ${ Number(margins.right) || 0 }px ${ Number(margins.bottom) || 0 }px ${ Number(margins.left) || 0 }px`;
}

function IntersectionObservers(options?: OPTIONS) {
  this.thresholds = (options && options.thresholds) ? options.thresholds : [0];
  this.initialRatio = (options && options.initialRatio) ? options.initialRatio : 0;
  this.selectAll = (options && options.selectAll) ? options.selectAll : false;
  this.callBack = () => {};
  this.observers = [];
}

function relative(root: Element | null, self, margins?: RECT) {
  const options = {
    root,
    rootMargin: margins ? buildMarginStr(margins) : '0px',
    threshold: self.thresholds,
  };
  self.observers.push(new IntersectionObserver((...args) => {
    self.callBack(...args);
  }, options));
}

IntersectionObservers.prototype.relativeTo = function (selector: string, margins?: RECT) {
  relative(document.querySelector(selector), this, margins);
  return this;
};

IntersectionObservers.prototype.relativeToViewport = function (margins?: RECT) {
  relative(null, this, margins);
  return this;
};

IntersectionObservers.prototype.observe = function (targetSelector: string, callback: (res: CallbackOptions) => any) {
  if (this.selectAll) {
    const targets = document.querySelectorAll(targetSelector);
    targets.forEach((target) => this.observers.forEach((observer) => observer.observe(target)));
  } else {
    const target = document.querySelector(targetSelector);
    this.observers.forEach((observer) => observer.observe(target));
  }
  this.callBack = (entries, observer) => {
    entries.forEach((entry) => {
      callback && typeof callback === 'function' && callback({
        boundingClientRect: entry.boundingClientRect,
        intersectionRatio: entry.intersectionRatio,
        intersectionRect: entry.intersectionRect,
        relativeRect: entry.rootBounds,
        time: entry.time,
      });
    });
  };
  return this;
};

IntersectionObservers.prototype.disconnect = function (targetSelector: string, callback: (res: CallbackOptions) => any) {
  this.callBack = () => {};
  this.observers = [];
  this.observers.forEach((observer) => {
    observer.disconnect();
  });
  return this;
};

const createIntersectionObserver: CreateIntersectionObserver = (options?, component?) => {
  return new IntersectionObservers(options);
};

export default createIntersectionObserver;
