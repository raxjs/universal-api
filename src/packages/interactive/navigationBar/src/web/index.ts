import setNavigationBarTitle from './setNavigationBarTitle';

export {
  setNavigationBarTitle,
};
export default {
  setNavigationBarTitle,
};
