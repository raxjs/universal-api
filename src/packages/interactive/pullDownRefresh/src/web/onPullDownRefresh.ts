import {
  normalizeSwitch,
} from '../common';
import {
  CONTAINER_NAME,
} from '@utils/constant';
import Events from '@utils/event';

class PullDownRefresh {
  enablePullDownRefresh: (params) => void;
  disablePullDownRefresh: () => void;
  private cb1: (params) => void;
  private cb2: (params) => void;
  private cb3: (params) => void;
  constructor() {
    const clsPrefix = '__universal_pulldownrefresh';

    const styles = {
      refresh: {
        position: 'relative',
        width: '100%',
        height: '50px',
        textAlign: 'center',
        display: 'flex',
        flexWrap: 'row',
        alignItems: 'center',
        justifyContent: 'center',
        zIndex: 9999,
      },
      refreshLoadingStyle: {
        height: '16px',
        width: '16px',
        marginRight: '10px',
        color: '#999',
      },
      refreshText: {
        fontSize: '14px',
        color: '#999',
      },
    };

    let refresh: HTMLElement | null = null;
    let refreshText: HTMLElement | null = null;
    let refreshLoadingImg: HTMLElement | null = null;
    /**
     * 获取刷新Dom
     * @param status {number} 1:下拉刷新状态 2:正在刷新状态
     * @returns
     */
    const _getRefresh = (status: number) => {
      refresh = document.getElementById(`${clsPrefix}_refresh`);
      refreshText = document.getElementById(`${clsPrefix}_refreshText`);
      refreshLoadingImg = document.getElementById(`${clsPrefix}_refreshLoadingImg`);
      if (!refresh) {
        refresh = document.createElement('div');
        refresh.id = `${clsPrefix}_refresh`;
        for (const key in styles.refresh) {
          if (Object.prototype.hasOwnProperty.call(styles.refresh, key)) {
            refresh.style[key] = styles.refresh[key];
          }
        }
      }
      if (!refreshText) {
        refreshText = document.createElement('div');
        refreshText.id = `${clsPrefix}_refreshText`;
        for (const key in styles.refreshText) {
          if (Object.prototype.hasOwnProperty.call(styles.refreshText, key)) {
            refreshText.style[key] = styles.refreshText[key];
          }
        }
      }
      if (!refreshLoadingImg) {
        refreshLoadingImg = document.createElement('img');
        refreshLoadingImg.id = `${clsPrefix}_refreshLoadingImg`;
        for (const key in styles.refreshLoadingStyle) {
          if (Object.prototype.hasOwnProperty.call(styles.refreshLoadingStyle, key)) {
            refreshLoadingImg.style[key] = styles.refreshLoadingStyle[key];
          }
        }
        refreshLoadingImg.setAttribute('src', 'https://gw.alicdn.com/imgextra/i4/O1CN01X5Adob1J0TGk79HNn_!!6000000000966-1-tps-400-400.gif');
      }
      refresh.appendChild(refreshText);
      switch (status) {
        case 1:
          if (refresh !== document.body.firstElementChild) {
            document.body.insertBefore(refresh, document.body.firstElementChild);
          }
          if (refreshLoadingImg === refresh.firstElementChild) {
            refresh.removeChild(refreshLoadingImg);
          }
          refreshText.innerText = '下拉刷新';
          break;
        case 2:
          if (refreshLoadingImg !== refresh.firstElementChild) {
            refresh.insertBefore(refreshLoadingImg, refresh.firstElementChild);
          }
          refreshText.innerText = '更新中...';
          break;
        default:
          break;
      }
      return refresh;
    };

    // let cb1: (params) => void; let cb2: (params) => void; let
    //   cb3: (params) => void;

    /**
     * 开启手动下拉
     * @param triggerDistance 触发'pulldownrefresh'所需的距离
     */
    const enablePullDownRefresh = (triggerDistance) => {
      const _element = document.body;
      let _startPos = 0;
      let _transitionHeight = 0;
      if (!this.cb1) {
        this.cb1 = function (e) {
          // console.log('初始位置：', e.touches[0].pageY);
          _startPos = e.touches[0].pageY;
        };
        _element.addEventListener('touchstart', this.cb1, true);
      }
      if (!this.cb2) {
        this.cb2 = function (e) {
          // console.log('当前位置：', e.touches[0].pageY);
          _transitionHeight = e.touches[0].pageY - _startPos;
          // console.log(_transitionHeight)
          if (_transitionHeight > 0 && _transitionHeight < triggerDistance) {
            _getRefresh(1);
          }
        };
        _element.addEventListener('touchmove', this.cb2, true);
      }
      if (!this.cb3) {
        this.cb3 = function (e) {
          if (_transitionHeight > triggerDistance) {
            _getRefresh(2);
            // console.log("触发更新", 'on');
            (window as any).events.emit('pulldownrefresh');
          }
          setTimeout(() => {
            _startPos = 0;
            _transitionHeight = 0;
          }, 20);
        };
        _element.addEventListener('touchend', this.cb3, true);
      }
    };

    /**
     * 关闭手动下拉
     */
    const disablePullDownRefresh = () => {
      const _element = document.body;
      if (this.cb1) {
        _element.removeEventListener('touchstart', this.cb1);
        this.cb1 = undefined;
      }
      if (this.cb2) {
        _element.removeEventListener('touchmove', this.cb2);
        this.cb2 = undefined;
      }
      if (this.cb3) {
        _element.removeEventListener('touchend', this.cb3);
        this.cb3 = undefined;
      }
    };

    this.enablePullDownRefresh = enablePullDownRefresh;
    this.disablePullDownRefresh = disablePullDownRefresh;
  }
}
/**
 * 开启手动下拉刷新
 */
const onPullDownRefresh = normalizeSwitch(({
  pullRefresh = true, triggerDistance = 90, eventCallback = () => {}, success = () => {}, fail = () => {}, complete = () => {},
}) => {
  try {
    if (!(window as any).events) {
      (window as any).events = new Events();
    }

    (window as any).events.register('pulldownrefresh', eventCallback);
    const pullDownRefresh = new PullDownRefresh();
    if (pullRefresh) {
      pullDownRefresh.enablePullDownRefresh(triggerDistance);
    } else {
      pullDownRefresh.disablePullDownRefresh();
    }

    success();
    complete();
  } catch (error) {
    fail();
    complete();
  }
}, CONTAINER_NAME.WEB);

export default onPullDownRefresh;
