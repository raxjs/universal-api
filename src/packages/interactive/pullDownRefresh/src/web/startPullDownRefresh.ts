import { normalizeStart } from '../common';
import { CONTAINER_NAME } from '@utils/constant';
import Events from '@utils/event';

const _startPullDownRefresh = () => {
  if (!(window as any).events) {
    (window as any).events = new Events();
  }
  const clsPrefix = '__universal_pulldownrefresh';

  const styles = {
    refresh: {
      position: 'relative',
      width: '100%',
      height: '50px',
      textAlign: 'center',
      display: 'flex',
      flexWrap: 'row',
      alignItems: 'center',
      justifyContent: 'center',
      zIndex: 9999,
    },
    refreshLoadingStyle: {
      height: '16px',
      width: '16px',
      marginRight: '10px',
      color: '#999',
    },
    refreshText: {
      fontSize: '14px',
      color: '#999',
    },
  };

  let refresh: HTMLElement | null = null;
  let refreshText: HTMLElement | null = null;
  let refreshLoadingImg: HTMLElement | null = null;

  // console.log("_startPullDownRefresh start");
  refresh = document.getElementById(`${clsPrefix}_refresh`);
  refreshText = document.getElementById(`${clsPrefix}_refreshText`);
  refreshLoadingImg = document.getElementById(`${clsPrefix}_refreshLoadingImg`);
  if (!refresh) {
    refresh = document.createElement('div');
    refresh.id = `${clsPrefix}_refresh`;
    for (const key in styles.refresh) {
      if (Object.prototype.hasOwnProperty.call(styles.refresh, key)) {
        refresh.style[key] = styles.refresh[key];
      }
    }
  }
  if (!refreshText) {
    refreshText = document.createElement('div');
    refreshText.id = `${clsPrefix}_refreshText`;
    for (const key in styles.refreshText) {
      if (Object.prototype.hasOwnProperty.call(styles.refreshText, key)) {
        refreshText.style[key] = styles.refreshText[key];
      }
    }
  }
  if (!refreshLoadingImg) {
    refreshLoadingImg = document.createElement('img');
    refreshLoadingImg.id = `${clsPrefix}_refreshLoadingImg`;
    for (const key in styles.refreshLoadingStyle) {
      if (Object.prototype.hasOwnProperty.call(styles.refreshLoadingStyle, key)) {
        refreshLoadingImg.style[key] = styles.refreshLoadingStyle[key];
      }
    }
    refreshLoadingImg.setAttribute('src', 'https://gw.alicdn.com/imgextra/i4/O1CN01X5Adob1J0TGk79HNn_!!6000000000966-1-tps-400-400.gif');
  }

  if (refresh !== document.body.firstElementChild) {
    document.body.insertBefore(refresh, document.body.firstElementChild);
  }
  refresh.appendChild(refreshText);
  if (refreshLoadingImg !== refresh.firstElementChild) {
    refresh.insertBefore(refreshLoadingImg, refresh.firstElementChild);
  }
  refreshText.innerText = '更新中...';
  // console.log("触发更新");
};

const startPullDownRefresh = normalizeStart(({ success = () => {}, fail = () => {}, complete = () => {} }) => {
  try {
    _startPullDownRefresh();

    (window as any).events.emit('pulldownrefresh');

    success();
    complete();
  } catch (error) {
    fail();
    complete();
  }
}, CONTAINER_NAME.WEB);

export default startPullDownRefresh;

