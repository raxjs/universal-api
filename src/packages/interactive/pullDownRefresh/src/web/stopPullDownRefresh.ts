import { normalizeStop } from '../common';
import { CONTAINER_NAME } from '@utils/constant';

const _stopPullDownRefresh = () => {
  // console.log("_stopPullDownRefresh  start");
  const clsPrefix = '__universal_pulldownrefresh';
  const refresh = document.getElementById(`${clsPrefix}_refresh`);
  if (refresh) {
    document.body.removeChild(refresh);
  }
};

const stopPullDownRefresh = normalizeStop(({ success = () => {}, fail = () => {}, complete = () => {} }) => {
  try {
    _stopPullDownRefresh();

    success();
    complete();
  } catch (error) {
    fail();
    complete();
  }
}, CONTAINER_NAME.WEB);

export default stopPullDownRefresh;
