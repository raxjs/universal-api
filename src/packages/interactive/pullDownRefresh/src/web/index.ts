import startPullDownRefresh from './startPullDownRefresh';
import stopPullDownRefresh from './stopPullDownRefresh';
import onPullDownRefresh from './onPullDownRefresh';

export {
  onPullDownRefresh,
  startPullDownRefresh,
  stopPullDownRefresh,
};
export default {
  onPullDownRefresh,
  startPullDownRefresh,
  stopPullDownRefresh,
};
