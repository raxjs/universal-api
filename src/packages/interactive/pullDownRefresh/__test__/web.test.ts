import { testWebAPI } from '@/utils/__test__/util';
import { DOMUtil } from '@/utils/__test__/web';

testWebAPI('pullDownRefresh', async () => {
  jest.useFakeTimers();
  const { startPullDownRefresh, stopPullDownRefresh, onPullDownRefresh } = require('../src/index');

  await startPullDownRefresh();
  expect(document.body).toMatchSnapshot();

  await stopPullDownRefresh();
  expect(document.body).toMatchSnapshot();

  const mockCallback = jest.fn();
  await onPullDownRefresh({
    triggerDistance: 100,
    eventCallback: mockCallback,
  });
  const domUtil = new DOMUtil(document.body);
  domUtil.trigger('touchstart', { touches: [{ pageY: 0 }] });
  jest.advanceTimersByTime(30);
  domUtil.trigger('touchmove', { touches: [{ pageY: 60 }] });
  jest.advanceTimersByTime(30);
  expect(document.body).toMatchSnapshot();
  domUtil.trigger('touchmove', { touches: [{ pageY: 120 }] });
  jest.advanceTimersByTime(30);
  expect(document.body).toMatchSnapshot();
  expect(mockCallback.mock.calls.length).toBe(0);

  domUtil.trigger('touchend', { touches: [{ pageY: 120 }] });
  jest.advanceTimersByTime(30);
  expect(document.body).toMatchSnapshot();
  expect(mockCallback.mock.calls.length).toBe(1);

  jest.useRealTimers();
});
