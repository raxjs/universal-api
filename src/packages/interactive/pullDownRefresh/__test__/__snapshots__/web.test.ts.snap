// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Platform API: pullDownRefresh Test container: web 1`] = `
<body>
  <div
    id="__universal_pulldownrefresh_refresh"
    style="position: relative; width: 100%; height: 50px; text-align: center; display: flex; flex-wrap: row; align-items: center; justify-content: center; z-index: 9999;"
  >
    <img
      id="__universal_pulldownrefresh_refreshLoadingImg"
      src="https://gw.alicdn.com/imgextra/i4/O1CN01X5Adob1J0TGk79HNn_!!6000000000966-1-tps-400-400.gif"
      style="height: 16px; width: 16px; margin-right: 10px; color: rgb(153, 153, 153);"
    />
    <div
      id="__universal_pulldownrefresh_refreshText"
      style="font-size: 14px; color: rgb(153, 153, 153);"
    >
      更新中...
    </div>
  </div>
</body>
`;

exports[`Test Platform API: pullDownRefresh Test container: web 2`] = `<body />`;

exports[`Test Platform API: pullDownRefresh Test container: web 3`] = `
<body>
  <div
    id="__universal_pulldownrefresh_refresh"
    style="position: relative; width: 100%; height: 50px; text-align: center; display: flex; flex-wrap: row; align-items: center; justify-content: center; z-index: 9999;"
  >
    <div
      id="__universal_pulldownrefresh_refreshText"
      style="font-size: 14px; color: rgb(153, 153, 153);"
    >
      下拉刷新
    </div>
  </div>
</body>
`;

exports[`Test Platform API: pullDownRefresh Test container: web 4`] = `
<body>
  <div
    id="__universal_pulldownrefresh_refresh"
    style="position: relative; width: 100%; height: 50px; text-align: center; display: flex; flex-wrap: row; align-items: center; justify-content: center; z-index: 9999;"
  >
    <div
      id="__universal_pulldownrefresh_refreshText"
      style="font-size: 14px; color: rgb(153, 153, 153);"
    >
      下拉刷新
    </div>
  </div>
</body>
`;

exports[`Test Platform API: pullDownRefresh Test container: web 5`] = `
<body>
  <div
    id="__universal_pulldownrefresh_refresh"
    style="position: relative; width: 100%; height: 50px; text-align: center; display: flex; flex-wrap: row; align-items: center; justify-content: center; z-index: 9999;"
  >
    <img
      id="__universal_pulldownrefresh_refreshLoadingImg"
      src="https://gw.alicdn.com/imgextra/i4/O1CN01X5Adob1J0TGk79HNn_!!6000000000966-1-tps-400-400.gif"
      style="height: 16px; width: 16px; margin-right: 10px; color: rgb(153, 153, 153);"
    />
    <div
      id="__universal_pulldownrefresh_refreshText"
      style="font-size: 14px; color: rgb(153, 153, 153);"
    >
      更新中...
    </div>
  </div>
</body>
`;
